crates/noirc_abi/src/lib.rs

```diff
@@ -100,13 +105,157 @@ impl Abi {
     pub fn num_parameters(&self) -> usize {
         self.parameters.len()
     }
+
+    /// Returns the number of field elements required to represent the ABI's input once encoded.
+    pub fn field_count(&self) -> u32 {
+        self.parameters.iter().map(|(_, typ)| typ.field_count()).sum()
+    }
+
     /// ABI with only the public parameters
     #[must_use]
     pub fn public_abi(self) -> Abi {
         let parameters: Vec<_> =
             self.parameters.into_iter().filter(|(_, param_type)| param_type.is_public()).collect();
         Abi { parameters }
     }
+
+    /// Encode a set of inputs as described in the ABI into a vector of `FieldElement`s.
+    pub fn encode(
+        self,
+        inputs: &BTreeMap<String, InputValue>,
+        allow_undefined_return: bool,
+    ) -> Result<Vec<FieldElement>, AbiError> {
+        let param_names = self.parameter_names();
+        let mut encoded_inputs = Vec::new();
+
+        for (param_name, param_type) in self.parameters.iter() {
+            let value = inputs
+                .get(param_name)
+                .ok_or_else(|| AbiError::MissingParam(param_name.clone()))?
+                .clone();
+
+            if !value.matches_abi(param_type) {
+                return Err(AbiError::TypeMismatch {
+                    param_name: param_name.to_string(),
+                    param_type: param_type.to_owned(),
+                    value,
+                });
+            }
+
+            // As the circuit calculates the return value in the process of calculating rest of the witnesses
+            // it's not absolutely necessary to provide them as inputs. We then tolerate an undefined value for
+            // the return value input and just skip it.
+            if allow_undefined_return
+                && param_name == MAIN_RETURN_NAME
+                && matches!(value, InputValue::Undefined)
+            {
+                let return_witness_len = param_type.field_count();
+
+                // We do not support undefined arrays for now - TODO
```

Comment:
Can you open up an issue for this TODO once merged?

Reply:
This is an existing TODO from when the inferred return value from main feature was first added IIRC

Reply:
Ahh okay, I'll open the issue for this once merged, as it was not Tom's addition